components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/OrientAsyncHelper.java

```diff
@@ -0,0 +1,203 @@
+  {
+    return asyncIterable(db, selectQuery, parameters, BUFFER_SIZE, TIMEOUT_SECONDS);
+  }
+
+  /**
+   * Executes with passed in {@link ODatabaseDocumentTx} connection the passed in {@link String} SELECT SQL command
+   * and returns a true asynchronous backed {@link Iterable} of {@link ODocument} instance, that can be later
+   * transformed into whatever caller needs. This means that this wrapper supports ANY SQL commands, even those
+   * performing aggregation on results, as long as iterator is not accumulated into a real array or list. Still, given
+   * the nature of non-blocking query, it's execution happens on a separate thread, hence, it will not see uncommited
+   * changes from the caller thread, if any. This method allows to tune buffer queue size and timeout.
+   *
+   * @param db             The {@link ODatabaseDocumentTx} database instance
+   * @param selectQuery    The SQL SELECT query in full
+   * @param parameters     The SQL SELECT named parameters (optional)
+   * @param bufferSize     The queue buffer size
+   * @param timeoutSeconds The timeout (in seconds) on queue handoff operations
+   */
+  public static Iterable<ODocument> asyncIterable(final ODatabaseDocumentTx db,
+                                                  final String selectQuery,
+                                                  @Nullable final Map<String, Object> parameters,
+                                                  final int bufferSize,
+                                                  final long timeoutSeconds)
+  {
+    checkNotNull(db);
+    checkNotNull(selectQuery);
+    checkArgument(bufferSize > 0);
+    checkArgument(timeoutSeconds >= 0);
+    final BlockingQueue<ODocument> queue = new ArrayBlockingQueue<>(bufferSize);
+
+    db.command(
+        new OSQLNonBlockingQuery<ODocument>(
+            selectQuery,
+            new QueueFeedingResultListener(timeoutSeconds, queue)
+        )
+    ).execute(parameters);
+
+    return new QueueConsumingIterable(timeoutSeconds, queue);
+  }
+
+  /**
+   * A "sentinel" {@link ODocument} instance that marks non blocking result end.
+   */
+  private static final ODocument SENTINEL = new ODocument();
+
+  /**
+   * The queue feeding {@link OCommandResultListener} implementation.
+   */
+  private static final class QueueFeedingResultListener
+      extends ComponentSupport
+      implements OCommandResultListener
+  {
+    private final long timeoutSeconds;
+
+    private final BlockingQueue<ODocument> queue;
+
+    public QueueFeedingResultListener(final long timeoutSeconds,
+                                      final BlockingQueue<ODocument> queue)
+    {
+      this.timeoutSeconds = timeoutSeconds;
+      this.queue = queue;
+    }
+
+    @Override
+    public boolean result(final Object o) {
+      final ODocument doc = (ODocument) o;
+      try {
+        queue.offer(doc, timeoutSeconds, TimeUnit.SECONDS);
```

Comment:
Does the use of 'offer' put the orient thread and the result-processing thread into a sort of race? If Orient is 10% faster at populating the queue than the processing thread is at processing documents, isn't the queue filling up and timing out inevitable (for a large enough result set)?  (I guess you could just set the queue to a very large size, but this seems like a complicated way of just loading all the data into memory before processing it.)